src/capped_hashmap.rs

```diff
@@ -0,0 +1,172 @@
+use std::{
+    cmp::Eq,
+    collections::{HashMap, VecDeque},
+    hash::Hash,
+};
+
+pub struct CappedHashMap<K, V>
+where
+    K: Hash + Eq + Copy + Clone,
+{
+    capacity: usize,
+    inner: HashMap<K, V>,
+    last_items: VecDeque<K>,
+}
+
+impl<K, V> CappedHashMap<K, V>
+where
+    K: Hash + Eq + Copy + Clone,
+{
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            // We add a little bit margin to the capacity, because we always insert first and then check
+            // if collection is full. So we don't want the hashmap to double it's capacity because of that.
+            capacity: capacity + 1,
+            inner: HashMap::with_capacity(capacity),
+            last_items: VecDeque::with_capacity(capacity),
+        }
+    }
+
+    /// Inserts an new item to the collection. Return Some(key) where key is the
+    /// key that was removed when we reach the max capacity. Otherwise returns None.
+    pub fn insert(&mut self, k: K, v: V) -> Option<K> {
```

Comment:
I don't think we should return the item that was removed, as this might be surprising behavior for something that is supposed to closely mimic a hashmap. If we want this API it should be named something else (`insert_and_get_removed_item` or something)

Reply:
> I don't think we should return the item that was removed

This is actually quite handy and it can help with us to things like this https://github.com/sigma0-xyz/zkbitcoin/issues/2#issuecomment-1902270914.

It's similar to what the core HashMap does with their `insert` fn. It returns the value of the replaced item and they don't call it something like `insert_and_get_replaced_item`.

I believe we can can this `add_entry` to avoid any confusion with the `HashMap::insert` fn.